### ISSUE TITLE
Rule proposal: `prefer-json-parse-buffer`

### ISSUE THREAD

**Issue by fisker** [sindresorhus/eslint-plugin-unicorn#1273]
When parsing a json file, there is no need to read it as string, because `JSON.parse` can parse `Buffer`.  [Real world case ](https://github.com/fisker/esm-utils/commit/04ea15c4dca43b8ebd6a9665ce34afaf5e86279d)

## Fail

```js
JSON.parse(await fs.readFile(file, 'utf8'));
```

## Pass

```js
JSON.parse(await fs.readFile(file));
```

```js
JSON.parse(await fs.readFile(file, 'gbk'));
```

This is also faster when use `fsPromises`, but this is bug, should not be the main reason for this rule. https://github.com/nodejs/node/issues/37583

**Comment by B2o5T** [comment]
I think is better use `require` than call `fs.readFile`
## Fail
``` js
JSON.parse(await fs.readFile(filePath, 'utf8'))
```
## Pass

``` js
const data = require(filePath)
```

**Comment by sindresorhus** [comment]
@B2o5T Your comment is not relevant to this issue. Also, the whole ecosystem is moving to ESM, which doesn't yet support JSON.

**Comment by silverwind** [comment]
> JSON.parse can parse Buffer

That seems like a non-standard extension `JSON.parse` of Node.js, so it may not work in other environments like Deno, but I guess `fs` will not be there either, so it's probably fine to match that specific case. It should work with destructured `readFile` and `readFileSync` too.

**Comment by fisker** [comment]
> That seems like a non-standard extension JSON.parse

I think it's not, `JSON.parse({[Symbol.toPrimitive](){return '[1]'}})`

**Comment by silverwind** [comment]
It does not generally seem to work with byte array:

```js
JSON.parse((new TextEncoder()).encode("{}"))
// Uncaught SyntaxError: Unexpected token , in JSON at position 3
```

Anyways, I think it's a good rule to have for a node-specific optimization.

**Comment by sindresorhus** [comment]
This is now accepted.

**Comment by zloirock** [comment]
@silverwind
> That seems like a non-standard extension `JSON.parse` of Node.js
> It does not generally seem to work with byte arrays

It works just because `JSON.parse` calls `ToString` inside (by the spec) and `Buffer#toString` converts the buffer to the string with `utf8` encoding by default. `TextEncoder#encode` returns just a typed array with `%TypedArray%#toString` that does not convert it to the original string.

**Comment by crystalfp** [comment]
Sorry to disagree, but typescript JSON.parse() does not accept anything but a string. See [microsoft/TypeScript #44944](https://github.com/microsoft/TypeScript/issues/44944). The only solution for me is to disable this rule.

**Comment by fisker** [comment]
I guess you'll have to disable in TS file.

**Comment by crystalfp** [comment]
Or better, in the `eslint.yaml` file where I added: `unicorn/prefer-json-parse-buffer: off` to the `rules:` section.

**Comment by devinrhode2** [comment]
I don't know why this is _preferable_. I assume the argument is simply: shorter code is preferable. But, it's forces (at least strongly encourages) developers to chase down this GH thread to realize `JSON.parse` calls `Buffer#toString` which uses `utf8`.

But... why not go a step further, if want short code, encourage using some sort of fs-extras `readJson` method?
(Example: [`await fs.readJson('./package.json')`](https://github.com/jprichardson/node-fs-extra/blob/HEAD/docs/readJson.md))

**Comment by fisker** [comment]
Mentioned in the description

> This is also faster when use fsPromises, but this is bug, should not be the main reason for this rule. https://github.com/nodejs/node/issues/37583

**Comment by fisker** [comment]
Main reason we are reading JSON files a lot is because we can't `import` JSON yet, compare to `require('./foo.json')` we used do, `JSON.parse(await fs.readFile(file, 'utf8'))` vs `JSON.parse(await fs.readFile(file))`, I prefer the shorter one.

**Comment by silverwind** [comment]
> we can't import JSON yet

JSON modules are unflagged since https://github.com/nodejs/node/pull/41736, fyi.

**Comment by fisker** [comment]
Wow~ But, without `assertion`? Surprised.

**Comment by addaleax** [comment]
Reading a file as a `Buffer` and then converting it to a string is *absolutely* worse than just reading it as a string, with worse CPU and memory performance. The only sensible lint rule is the exact opposite of what is being enforced here.

**Comment by devinrhode2** [comment]
I'm content with this being the main reason for this rule:
> This is also faster when use fsPromises, but this is bug, should not be the main reason for this rule. https://github.com/nodejs/node/issues/37583

@addaleax do you mind providing some insight why `Buffer`->`string` is worse for CPU and memory?

A benchmark could assume json files are less than 10 pages, where a page is around 100 lines.

**Comment by fisker** [comment]
Maybe we should disable this rule in `recommend`?

**Comment by privatenumber** [comment]
@fisker Should I open an new issue to request `prefer-json-parse-buffer` to be moved out of `recommended`?

**Comment by fisker** [comment]
Send a PR, I'll approve.
